File: content/04/code/ruby/02-rule-keyword-added/lib/shouty.rb
module Shouty
  class Person
    attr_reader :messages_heard
    
    def initialize(network)
      @messages_heard = []
      @network        = network
      
      @network.subscribe(self)
    end

    def move_to(distance)
    end

    def shout(message)
      @network.broadcast(message);
    end

    def hear(message)
      @messages_heard << message
    end
  end

  class Network
    def initialize
      @listeners = []
    end

    def subscribe(person)
      @listeners << person
    end

    def broadcast(message)
      @listeners.each do |listener|
        listener.hear(message)
      end
    end
  end
end

File: content/bdd-with-cucumber/04/code/js/11-replace-hardcoded-lucy-instance/src/shouty.js
class Person {
  constructor(network) {
    this.messages = []
    this.network      = network

    this.network.subscribe(this)
  }
  
  shout(message) {
    this.network.broadcast(message)
  }

  hear(message) {
    this.messages.push(message)
  }

  messagesHeard() {
    return this.messages
  }
}

class Network {
  constructor() {
    this.listeners = []
  }

  subscribe(person) {
    this.listeners.push(person)
  }

  broadcast(message) {
    this.listeners.forEach(listener => { listener.hear(message) })
  }
}

module.exports = {
  Person  : Person,
  Network : Network
}
